text/0000-glimmer-components.md

```diff
@@ -0,0 +1,1022 @@
+
+### Element Modifiers
+
+Glimmer components have outer HTML semantics, which means that whatever is put
+in the template is what we see in the final render. This also means that they
+can have _multiple_ root elements:
+
+```hbs
+<div ...attributes>
+  First root element
+</div>
+<div>
+  Second root element
+</div>
+```
+
+It is also possible that there are _no_ root elements:
+
+```hbs
+This is a text only component: {{@foo}}
+```
+
+This makes the `element` property of classic components awkward, since there is
+no one single referenceable element. A previous proposal for solving this
+problem was a `bounds` property which contained `firstNode` and `lastNode`, but
+this could be very confusing to users since these nodes would likely have to be
+text nodes (how Glimmer tracks component boundaries internally), and would force
+users to write unintuitive looping logic to traverse nodes:
+
+```js
+let { firstNode, lastNode } = bounds;
+let currentNode = firstNode;
+
+while (currentNode !== lastNode) {
+  // do something with current node
+  currentNode = currentNode.nextSibling;
+}
+```
+
+This exposes users to low level APIs which are fairly complicated for most
+simple use cases, and can be error prone. We could also attempt to add an
+`element` property for cases where templates only have one root level element,
+but this represents a refactoring hazard if users attempt to rewrite code, and
+means we have to deal with a few complexities (what if a component has top level
+text nodes?)
+
+Instead, we can use new element modifiers which allow users to run code
+declaratively on the element:
+
+```hbs
+<div {{did-render this.setupElement}} ...attributes>
+  First root element
+</div>
+<div {{did-render this.setupElement}}>
+  Second root element
+</div>
+```
+
+This allows users to target which elements they care about directly. It also
+allows for code to be run on elements which are contained in conditionals,
+meaning users do not need to add messy logic to track component state
+themselves:
+
+```hbs
+{{#if shouldRender}}
+  <div
+    {{did-render this.setupElement}}
+    {{will-destroy this.teardownElement}}
+  >
+
+  </div>
+{{/if}}
+```
+
+Finally, it means that users can bypass component code altogether by using
+helpers:
+
+```hbs
+<div {{did-render setScrollPosition @scrollPosition}}></div>
+```
+
+This RFC proposes adding two element modifiers: `{{did-render}}` and
+`{{will-destroy}}`.
+
+#### `{{did-render}}`
+
+This modifier runs immediately after the element is inserted, and again whenever
+any of the arguments passed to it update, including the function passed as the
+first argument. It has the following timing semantics:
+
+* **Always**
+  * called _after_ the enclosing component's `didUpdateArgs` hook
+  * called _after_ any child element's `{{did-render}}` modifiers
+  * called in definition order in the template
+* **May or May Not**
+  * be called in the same tick as DOM insertion
```

Comment:
Seems tricky to say this isn't guaranteed in practice, since changing it later will almost certainly break app code, right?

Reply:
Relatedly, there are different use cases for accessing a DOM element before layout and after layout. If you want to do some kind of setup or initialization (e.g. a jQuery UI control), you want to do it before layout/paint to ensure there's no ugly flash of empty content. But if you want to do some sort of measurement (some behavior based on the element's natural width or height, for example) you want to do this after layout.

Today people typically have to learn about `afterRender` if they want the post-insertion, post-layout element. Would this be better modeled as a different element modifier/lifecycle hook?

Reply:
@tomdale if what you are doing is based on the DOM, you should be using the element modifiers which _do_ give you the guarantees that you are mentioning...

Reply:
I don't think there is anything to debate here... we merged the RFC that specified the timing of this.

The modifiers are installed pre-layout. This is why I'm unsure about calling this `{{did-render}}`. The modifier installation is [scheduled](https://github.com/glimmerjs/glimmer-vm/blob/e838d9dd02c41d4c57c14e47ac4ce5f5d89a48dc/packages/%40glimmer/runtime/lib/environment.ts#L192-L198) and is called from bottom-up in the commit of the render transaction. Because of this I'm not sure if `{{did-insert}}` is any better because you "inserting the modifier" not the element. 

If you wanted a modifier to measure something you would need to do and `afterRender` in the `installModifier` hook.